src/library/autodj/autodjprocessor.cpp

```diff
@@ -817,61 +965,204 @@ void AutoDJProcessor::calculateTransition(DeckAttributes* pFromDeck,
+        // Assume a zero length outro.
+        // The outroEnd is automatically placed by AnalyzerSilence, so use
+        // that as a fallback if the user has not placed outroStart.
+        if (outroEnd >= 0) {
+            outroStart = outroEnd;
+        } else {
+            outroEnd = fromTrackDuration;
         }
+    }
+    double outroLength = outroEnd - outroStart;
 
-        // Guard against tracks shorter than the desired transition period.
-        // Use a sensible cap.
-        double nextTransitionTime = math_min3(desiredTransitionTime,
-                                              fromTrackDuration / 2,
-                                              toTrackDuration / 2);
+    double introStart;
+    if (seekToStartPoint) {
+        introStart = getIntroStartPosition(pToDeck);
+    } else {
+        introStart = pToDeck->playPosition() * toTrackDuration;
+    }
 
-        double fromTrackStartFadeOut = fromTrackOutroEnd - nextTransitionTime;
+    double introEnd = getIntroEndPosition(pToDeck);
+    if (introEnd < introStart) {
+        // introEnd is invalid. Assume a zero length intro.
+        // The introStart is automatically placed by AnalyzerSilence, so use
+        // that as a fallback if the user has not placed introEnd.
+        introEnd = introStart;
+    }
 
-        if (fromTrackDuration > 0.0) {
-            pFromDeck->posThreshold = fromTrackStartFadeOut / fromTrackDuration;
-            pFromDeck->fadeDuration = nextTransitionTime / fromTrackDuration;
+    double introLength = introEnd - introStart;
+
+    switch (m_transitionMode) {
+    case TransitionMode::FullIntroOutro:
+        // Use the outro or intro length for the transition time, whichever is
+        // shorter. Let the full outro and intro play; do not cut off any part
+        // of either.
+        //
+        // In the diagrams below,
+        // - is part of a track outside the outro/intro,
+        // o is part of the outro
+        // i is part of the intro
+        // | marks the boundaries of the transition
+        //
+        // When outro > intro:
+        // ------ooo|ooo|
+        //          |iii|------
+        //
+        // When outro < intro:
+        // ------|ooo|
+        //       |iii|iii-----
+        //
+        // If only the outro or intro length is marked but not both, use the one
+        // that is marked for the transition time. If neither is marked, fall
+        // back to the transition time from the spinbox.
+        if (outroLength > 0 && introLength > 0) {
+            if (outroLength > introLength) {
+                pFromDeck->fadeBeginPos = outroEnd - introLength;
+                pFromDeck->fadeEndPos = outroEnd;
+                pToDeck->startPos = introStart;
+            } else {
+                pFromDeck->fadeBeginPos = outroStart;
+                pFromDeck->fadeEndPos = outroEnd;
+                pToDeck->startPos = introStart;
+            }
+        } else if (outroLength > 0 && introLength <= 0) {
```

Comment:
you can omit "&& introLength <= 0" this clarifies the else case below.

Reply:
Yes, logically it can be omitted. But IMO it is easier to think about having both conditions explicitly specified.

Reply:
I see no problems from clang-tidy, clazy, nor cppcheck, so I see no reason to change this. I prefer code to be self explaining rather than require comments as much as possible.